internal/exec/stages/disks/luks.go

```diff
@@ -531,3 +559,87 @@ func (d LuksDump) hasFlag(flag string) bool {
 	}
 	return false
 }
+
+// collect the cex card domains xx.xxxx,
+func cexCardDomain() (string, error) {
+	var ret []string
+	domain, err := exec.Command(distro.LszcryptCmd(), "-V", "--ccaonly", "--queueonly").CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("lszcrypt: No crypto card devices found: %w", err)
+	}
+	// Skip the unwanted lines before the actual value
+	col := strings.Split(string(domain), "\n")[2:]
+	// Pattern to match the first column to grep one or more domains without card number.
+	for _, line := range col {
+		match := cexRegx.FindStringSubmatch(line)
+		if match != nil {
+			ret = append(ret, match[1])
+		}
+	}
+	return strings.Join(ret, ","), nil
+}
+
+// Arguments specific to zkey generation
+func zkeySecGenArgs(luks types.Luks) []string {
+	ret := []string{
+		"generate",
+		"--name",
+		"root_xts",
+		"--xts",
+		"--volumes",
+		execUtil.DeviceAlias(*luks.Device),
+		"--description",
+		"root Masterkey",
+	}
+	// Verify the Keytype for securekey
+	if !util.NilOrEmpty(luks.Cexmode.Keytype) {
+		ret = append(ret, "--key-type", *luks.Cexmode.Keytype)
+	} else {
+		ret = append(ret, "--key-type", "CCA-AESCIPHER")
+	}
+	// Sector size are optional, if not provided 512
+	if luks.Cexmode.Sectorsize != nil {
+		ret = append(ret, "--sector-size", "4096")
+	}
+	return ret
+}
+
+// Secure key Generation using zkey.
+func zkeySecKeyGen(luks types.Luks) error {
+	args := zkeySecGenArgs(luks)
+	dom, err := cexCardDomain()
+	if err != nil {
+		return fmt.Errorf("Error: %w", err)
+	}
+	args = append(args, "--apqns", dom)
+	_, err = exec.Command(distro.ZkeyCmd(), args...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Error generating securekey: %w", err)
+	}
+	return nil
+}
+
+// zkey Cryptsetup args generate pbkdf algo as argon2i.
+// fips does not support argon2i. hence adding the
+// args manually by verifying the keysize and keytype.
+func zkeySecCryptGenArg(luks types.Luks) []string {
+	ret := []string{
+		"--master-key-file",
+		"/etc/zkey/repository/root_xts.skey",
+	}
+	if !util.NilOrEmpty(luks.Cexmode.Keytype) && *luks.Cexmode.Keytype == "CCA-AESDATA" {
+		ret = append(ret, "--key-size", "1024")
+	} else {
+		ret = append(ret, "--key-size", "2176")
+	}
```

Comment:
Do I understand correctly that this arg doesn't actually control the key size since we're explicitly passing in the volume key file to use, but that it's just for checking its size?

Reply:
As far as i understand , the `cryptsetup` wants  `--key-size` when we use the `--volume-key-file` in xts mode.  if we do not provide key size the cryptsetup somehow use default 2*256 bits from the file  while formatting the disk and set the key size to 512bits (64bytes). 

 Without `--key-size`  somehow it  wont do the protected key mechanism with CEX card and no longer protected;  The volume can be open with passkey without CEX card . Also The zkey validation will fail as well as it fails to set the verification pattern  in the Token.

```
zkey-cryptsetup: # Device /dev/dasdb1 READ lock released.

zkey-cryptsetup: # PBKDF pbkdf2-sha256, time_ms 2000 (iterations 0).

zkey-cryptsetup: # PBKDF pbkdf2-sha256, time_ms 2000 (iterations 0).

zkey-cryptsetup: Installing SIGINT/SIGTERM handler
zkey-cryptsetup: Volume key size: 64
zkey-cryptsetup: Integrity: 'none'
zkey-cryptsetup: Integrity key size: 0
zkey-cryptsetup: The volume key size 64 is not valid for the cipher mode 'xts-plain64'
zkey-cryptsetup: # Releasing crypt device /dev/dasdb1 context.

zkey-cryptsetup: # Releasing device-mapper backend.

zkey-cryptsetup: # Closing read only fd for /dev/dasdb1.
```

eg:
I tried in one of the OCP worker node by formatting the disk as per the  `--key-size`   and the `cryptsetup luksDump` show the key size exactly as the key-size we provided.

```
sh-5.1# cryptsetup luksFormat /dev/dasdb1 --pbkdf pbkdf2 --volume-key-file '/etc/zkey/repository/secure_AESCipherKey4.skey' --key-size 2176 --cipher paes-xts-plain64 
WARNING: Device /dev/dasdb1 already contains a 'crypto_LUKS' superblock signature.

WARNING!
========
This will overwrite data on /dev/dasdb1 irrevocably.

Are you sure? (Type 'yes' in capital letters): YES
Enter passphrase for /dev/dasdb1: 
Verify passphrase: 
sh-5.1# cryptsetup luksDump /dev/dasdb1
LUKS header information
Version:       	2
Epoch:         	3
Metadata area: 	16384 [bytes]
Keyslots area: 	16744448 [bytes]
UUID:          	62910b38-1e69-4ed0-9e13-b641c423d0a0
Label:         	(no label)
Subsystem:     	(no subsystem)
Flags:       	(no flags)

Data segments:
  0: crypt
	offset: 16777216 [bytes]
	length: (whole device)
	cipher: paes-xts-plain64
	sector: 4096 [bytes]

Keyslots:
  0: luks2
	Key:        2176 bits
	Priority:   normal
	Cipher:     aes-xts-plain64
	Cipher key: 512 bits
	PBKDF:      pbkdf2
	Hash:       sha256
	Iterations: 1865792
	Salt:       cf ac 97 b8 49 92 cf 53 33 85 7e 7c 5e b1 bc 48 
	            5d 1a 3d f9 39 46 16 b8 3c 7f 84 7d 72 e7 64 78 
	AF stripes: 4000
	AF hash:    sha256
	Area offset:32768 [bytes]
	Area length:1089536 [bytes]
	Digest ID:  0
```

However we do not provide the `--key-size`  here is the output. 

```
sh-5.1# cryptsetup luksFormat /dev/dasdb1 --pbkdf pbkdf2 --volume-key-file '/etc/zkey/repository/secure_AESCipherKey5.skey'  --cipher paes-xts-plain64 
WARNING: Device /dev/dasdb1 already contains a 'crypto_LUKS' superblock signature.

WARNING!
========
This will overwrite data on /dev/dasdb1 irrevocably.

Are you sure? (Type 'yes' in capital letters): YES
Enter passphrase for /dev/dasdb1: 
Verify passphrase: 
sh-5.1# cryptsetup luksDump /dev/dasdb1
LUKS header information
Version:       	2
Epoch:         	3
Metadata area: 	16384 [bytes]
Keyslots area: 	16744448 [bytes]
UUID:          	473b0977-45ff-4194-8b53-36dc4babdca1
Label:         	(no label)
Subsystem:     	(no subsystem)
Flags:       	(no flags)

Data segments:
  0: crypt
	offset: 16777216 [bytes]
	length: (whole device)
	cipher: paes-xts-plain64
	sector: 4096 [bytes]

Keyslots:
  0: luks2
	Key:        512 bits
	Priority:   normal
	Cipher:     aes-xts-plain64
	Cipher key: 512 bits
	PBKDF:      pbkdf2
	Hash:       sha256
	Iterations: 1804776
	Salt:       f0 d2 15 e9 06 ad e6 7b a8 3b 28 38 7f 0b cb 99 
	            1e 65 4f 49 58 2d 63 67 3a 5f 1e 2e 2d e6 d0 b3 
	AF stripes: 4000
	AF hash:    sha256
	Area offset:32768 [bytes]
	Area length:258048 [bytes]
	Digest ID:  0
```

Reply:
@jlebon  adding on above ^ ,  I had word with zcrypto team.  Here is the explanation for that.

```
If you pass a key to paes that is a valid clear key size, then it uses it as clear key.
So in your case it reads the first 64 bytes from the secure key file and uses them as clear key value for AES-XTS. And for this it of course does not need any crypto cards.
The reason why this is done that way, is because all kernel ciphers need to run a selftest when they are loaded, and this self test can only be performed with clear keys. So paes must also accept clear key as key. Valid secure keys will never have a size that matches a valid clear key size.
So you must be careful and always specify the correct key size during luksFormat.
```
So the key-size must be verified and use that size as `--key-size` for  luksFormat, 
So in the ignition  we could  read the secure key file-size  which guarantees  the key-size.
Kindly suggest.

Reply:
Gotcha, thanks. That clears it up nicely.

I think one confusing thing here is that we're hardcoding key lengths instead of using the source of truth, which is the key that `zkey generate` generated. Isn't just the size of the key file itself?

Reply:
Yes,  it is the size of the key file. The following is key-type CCA-AESCIPHER which generates 272bytes size. CCA-AESDATA generates 128bytes size.  Usual cases `zkey` command manages all argument including the luksFormat and verification. However zkey generates the command arg `pbdkf` algo as `argon2i` where FIPS does not support.

```
sh-5.1# cd /etc/zkey/repository/
sh-5.1# ls
secure_AESCipherKey5.info  secure_AESCipherKey5.skey
sh-5.1# ls -l
total 8
-rw-r-----. 1 root zkeyadm 321 Mar  5 04:26 secure_AESCipherKey5.info
-rw-r-----. 1 root zkeyadm 272 Mar  5 04:26 secure_AESCipherKey5.skey
sh-5.1# KEYBITSIZE=$(stat --printf="%s" secure_AESCipherKey5.skey)
sh-5.1# KEYBITSIZE=$((${KEYBITSIZE}*8))
sh-5.1# echo $KEYBITSIZE
2176
```

Reply:
Then I'd say instead of hardcoding these values, let's just `os.Stat` the keyfile and use its length (* 8) as the argument.